dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/openapi/ApiFinalise.java

```diff
@@ -473,24 +475,106 @@ private static String getSharedNameForDeclaringType(Api.Parameter p) {
               endpoints.forEach(
                   e ->
                       e.getMethods()
                           .forEach(
                               method ->
-                                  endpointByMethod.compute(
-                                      method, (k, v) -> mergeEndpoints(v, e, method))));
-              api.getEndpoints().put(path, endpointByMethod);
+                                  endpointByMethod
+                                      .computeIfAbsent(
+                                          method, (Function) o -> new ArrayList<Api.Endpoint>())
+                                      .add(e)));
+
+              for (Map.Entry<RequestMethod, List<Api.Endpoint>> methodAndEndpoints :
+                  endpointByMethod.entrySet()) {
+                toPathsAndOperations(methodAndEndpoints, path);
+              }
             });
   }
 
+  private void toPathsAndOperations(
+      Map.Entry<RequestMethod, List<Api.Endpoint>> methodAndEndpoints, String path) {
+    if (methodAndEndpoints.getValue().size() > 1) {
+      Api.Endpoint nonFragmentEndpoint = getPreferredEndpoint(methodAndEndpoints);
+
+      for (int i = 0; i < methodAndEndpoints.getValue().size(); i++) {
+        if (!methodAndEndpoints.getValue().get(i).equals(nonFragmentEndpoint)) {
+          Api.Endpoint endpoint = methodAndEndpoints.getValue().get(i);
+          if (isMergeable(endpoint, nonFragmentEndpoint)) {
+            nonFragmentEndpoint =
+                mergeEndpoints(nonFragmentEndpoint, endpoint, methodAndEndpoints.getKey());
+          } else {
+            AbstractMap.SimpleEntry<String, EnumMap<RequestMethod, Api.Endpoint>>
+                fragmentPathAndOperation = fragmentise(methodAndEndpoints.getKey(), endpoint, path);
+            api.getEndpoints()
+                .put(fragmentPathAndOperation.getKey(), fragmentPathAndOperation.getValue());
+          }
+        }
+      }
+      api.getEndpoints()
+          .computeIfAbsent(path, s -> new EnumMap<>(RequestMethod.class))
+          .put(methodAndEndpoints.getKey(), nonFragmentEndpoint);
+    } else {
+      api.getEndpoints()
+          .computeIfAbsent(path, s -> new EnumMap<>(RequestMethod.class))
+          .put(methodAndEndpoints.getKey(), methodAndEndpoints.getValue().get(0));
+    }
+  }
+
+  private Api.Endpoint getPreferredEndpoint(
+      Map.Entry<RequestMethod, List<Api.Endpoint>> methodAndEndpoints) {
+    for (int i = 0; i < methodAndEndpoints.getValue().size(); i++) {
+      if (consumesOrProducesJson(methodAndEndpoints.getValue().get(i))) {
+        return methodAndEndpoints.getValue().get(i);
+      }
+    }
+
+    return methodAndEndpoints.getValue().get(0);
+  }
+
+  private AbstractMap.SimpleEntry<String, EnumMap<RequestMethod, Api.Endpoint>> fragmentise(
+      RequestMethod requestMethod, Api.Endpoint endpoint, String path) {
+    EnumMap<RequestMethod, Api.Endpoint> requestMethodObjectEnumMap =
```

Comment:
`requestMethodObjectEnumMap` is a bit unnecessary long and hard to read. we do not have to repeat information in a name that is apparent from the type. Use names to add information the types do not provide. If we don't have more to say make it as short as possible without being confusing. E.g. here `endpoints` would be fine. Also feel free to use `Map.of(x,y)` even for places where we could use a `EnumMap`. It is just so much more readable and for a single entry map there is no difference in cost.